From 81987d968ab831c08bd7e42a46f4a4823729cf1e Mon Sep 17 00:00:00 2001
From: Collin Funk <collin.funk1@gmail.com>
Date: Thu, 16 Jul 2026 19:31:57 -0700
Subject: [PATCH] talkd: avoid a buffer overflow with large DNS names

Reported privately by Tristan Madani <tristan@talencesecurity.com>.

* talkd/announce.c: Include intprops.h and inttostr.h.
(N_LINES, N_CHARS): Remove macros.
(LINE): Remove type.
(format_line, finish_line): Remove functions.
(print_mesg): Allocate enough bytes for the announcement.
* talkd/talkd.c (hostname_len): New variable.
(talkd_init): Initialize it.
* talkd/intalkd.h: Include idx.h
(hostname_len): Add declaration.
---
 talkd/announce.c | 199 +++++++++++++++++++++++++++++------------------
 talkd/intalkd.h  |   3 +
 talkd/talkd.c    |   2 +
 3 files changed, 127 insertions(+), 77 deletions(-)

diff --git a/talkd/announce.c b/talkd/announce.c
index 184d15c7..f5d71476 100644
--- a/talkd/announce.c
+++ b/talkd/announce.c
@@ -24,99 +24,144 @@
 
 #undef MAX
 #define MAX(a,b) ( (a) > (b) ? (a) : (b) )
-#define N_LINES 5
-#define N_CHARS 256
 
 #include <libinetutils.h>
 
-typedef struct
-{
-  int ind;
-  int max_size;
-  char line[N_LINES][N_CHARS];
-  int size[N_LINES];
-  char buf[N_LINES * N_CHARS + 3];
-} LINE;
-
-static void
-init_line (LINE *lp)
-{
-  memset (lp, 0, sizeof *lp);
-}
-
-static void
-format_line (LINE *lp, const char *fmt, ...)
-{
-  va_list ap;
-  int i = lp->ind;
-
-  if (lp->ind >= N_LINES)
-    return;
-  lp->ind++;
-  va_start (ap, fmt);
-  lp->size[i] = vsnprintf (lp->line[i], sizeof lp->line[i], fmt, ap);
-  lp->max_size = MAX (lp->max_size, lp->size[i]);
-  va_end (ap);
-}
-
-static char *
-finish_line (LINE *lp)
-{
-  int i;
-  char *p;
-
-  p = lp->buf;
-  *p++ = '\a';
-  *p++ = '\r';
-  *p++ = '\n';
-  for (i = 0; i < lp->ind; i++)
-    {
-      char *q;
-      int j;
-
-      for (q = lp->line[i]; *q; q++)
-	*p++ = *q;
-      for (j = lp->size[i]; j < lp->max_size + 2; j++)
-	*p++ = ' ';
-      *p++ = '\r';
-      *p++ = '\n';
-    }
-  *p = 0;
-  return lp->buf;
-}
+#include "intprops.h"
+#include "inttostr.h"
 
 static int
 print_mesg (char *tty, CTL_MSG *request, char *remote_machine)
 {
   time_t t;
-  LINE ln;
-  char *buf;
-  struct tm *tm;
-  struct iovec iovec;
-  char *cp;
-
   time (&t);
-  tm = localtime (&t);
-  init_line (&ln);
-  format_line (&ln, "");
-  format_line (&ln, "Message from Talk_Daemon@%s at %d:%02d ...",
-	       hostname, tm->tm_hour, tm->tm_min);
-  format_line (&ln, "talk: connection requested by %s@%s",
-	       request->l_name, remote_machine);
-  format_line (&ln, "talk: respond with:  talk %s@%s",
-	       request->l_name, remote_machine);
-  format_line (&ln, "");
-  format_line (&ln, "");
-  buf = finish_line (&ln);
+  struct tm *tm = localtime (&t);
 
-  iovec.iov_base = buf;
-  iovec.iov_len = strlen (buf);
+  char tm_hour_buf[INT_BUFSIZE_BOUND (int)];
+  char const *tm_hour_str = inttostr (tm->tm_hour, tm_hour_buf);
+  idx_t tm_hour_len = strlen (tm_hour_str);
 
+  char tm_min_buf[INT_BUFSIZE_BOUND (int)];
+  char *tm_min_str = inttostr (tm->tm_min, tm_min_buf);
+  if (0 <= tm->tm_min && tm->tm_min < 10)
+    *--tm_min_str = '0';
+  idx_t tm_min_len = strlen (tm_min_str);
+
+  /* "Message from Talk_Daemon@%s at %d:%02d ..."  */
+  static const char line1_prefix[] = "Message from Talk_Daemon@";
+  static const char line1_part2[] = " at ";
+  static const char line1_part3[] = " ...";
+
+  /* Don't subtract the NUL byte from LINE1_PART to make up for the
+     ':' character.  */
+  idx_t line1_len = ((sizeof line1_prefix - 1) + hostname_len
+		     + sizeof line1_part2 + tm_hour_len + tm_min_len
+		     + (sizeof line1_part3 - 1));
+
+  idx_t l_name_len = strlen (request->l_name);
+  idx_t remote_machine_len = strlen (remote_machine);
+
+  /* "talk: connection requested by %s@%s"  */
+  static const char line2_prefix[] = "talk: connection requested by ";
+
+  /* Don't subtract the NUL byte to make up for the '@' character.  */
+  idx_t line2_len = sizeof line2_prefix + l_name_len + remote_machine_len;
+
+  /* "talk: respond with:  talk %s@%s"  */
+  static const char line3_prefix[] = "talk: respond with:  talk ";
+
+  /* Don't subtract the NUL byte to make up for the '@' character.  */
+  idx_t line3_len = sizeof line3_prefix + l_name_len + remote_machine_len;
+
+  /* Get the maximum line length.  */
+  idx_t max_line_len = MAX (line1_len, MAX (line2_len, line3_len));
+
+  /* Begin with an alarm followed by CRLF.  Each of the 5 lines is filled
+     to MAX_LINE_LEN + 2 with spaces followed by a CRLF.  */
+  idx_t size = ((max_line_len + 2 + 2) * 5) + 3;
+  char *buf = malloc (size);
+  if (!buf)
+    {
+      syslog (LOG_ERR, "Out of memory");
+      exit (EXIT_FAILURE);
+    }
+
+  char *p = buf;
+
+  *p++ = '\a';
+  *p++ = '\r';
+  *p++ = '\n';
+
+  /* Line  0.  */
+  memset (p, ' ', max_line_len + 2);
+  p += max_line_len + 2;
+  *p++ = '\r';
+  *p++ = '\n';
+
+  /* Line 1.  */
+  memcpy (p, line1_prefix, sizeof line1_prefix - 1);
+  p += sizeof line1_prefix - 1;
+  memcpy (p, hostname, hostname_len);
+  p += hostname_len;
+  memcpy (p, line1_part2, sizeof line1_part2 - 1);
+  p += sizeof line1_part2 - 1;
+  memcpy (p, tm_hour_str, tm_hour_len);
+  p += tm_hour_len;
+  *p++ = ':';
+  memcpy (p, tm_min_str, tm_min_len);
+  p += tm_min_len;
+  memcpy (p, line1_part3, sizeof line1_part3 - 1);
+  p += sizeof line1_part3 - 1;
+  idx_t line1_spaces = (max_line_len + 2) - line1_len;
+  memset (p, ' ', line1_spaces);
+  p += line1_spaces;
+  *p++ = '\r';
+  *p++ = '\n';
+
+  /* Line 2.  */
+  memcpy (p, line2_prefix, sizeof line2_prefix - 1);
+  p += sizeof line2_prefix - 1;
+  memcpy (p, request->l_name, l_name_len);
+  p += l_name_len;
+  *p++ = '@';
+  memcpy (p, remote_machine, remote_machine_len);
+  p += remote_machine_len;
+  idx_t line2_spaces = (max_line_len + 2) - line2_len;
+  memset (p, ' ', line2_spaces);
+  p += line2_spaces;
+  *p++ = '\r';
+  *p++ = '\n';
+  memcpy (p, line3_prefix, sizeof line3_prefix - 1);
+  p += sizeof line3_prefix - 1;
+  memcpy (p, request->l_name, l_name_len);
+  p += l_name_len;
+  *p++ = '@';
+  memcpy (p, remote_machine, remote_machine_len);
+  p += remote_machine_len;
+
+  /* Line 3.  */
+  idx_t line3_spaces = (max_line_len + 2) - line3_len;
+  memset (p, ' ', line3_spaces);
+  p += line3_spaces;
+  *p++ = '\r';
+  *p++ = '\n';
+
+  /* Line 4.  */
+  memset (p, ' ', max_line_len + 2);
+  p += max_line_len + 2;
+  *p++ = '\r';
+  *p++ = '\n';
+
+  struct iovec iovec = {.iov_base = buf,.iov_len = size };
+
+  char *cp;
   if ((cp = inetutils_ttymsg (&iovec, 1, tty, RING_WAIT - 5)) != NULL)
     {
       syslog (LOG_ERR, "%s", cp);
+      free (buf);
       return FAILED;
     }
+  free (buf);
   return SUCCESS;
 }
 
diff --git a/talkd/intalkd.h b/talkd/intalkd.h
index 8ac32d44..692b7f1f 100644
--- a/talkd/intalkd.h
+++ b/talkd/intalkd.h
@@ -32,6 +32,8 @@
 #include <unistd.h>
 #include <errno.h>
 
+#include "idx.h"
+
 #ifndef INADDR_NONE
 # define INADDR_NONE -1
 #endif
@@ -48,6 +50,7 @@ extern unsigned int timeout;
 extern time_t max_idle_time;
 extern time_t max_request_ttl;
 extern char *hostname;
+extern idx_t hostname_len;
 
 #define os2sin_addr(cp) (((struct sockaddr_in *)&(cp))->sin_addr)
 
diff --git a/talkd/talkd.c b/talkd/talkd.c
index 2c97810b..3ab2cc04 100644
--- a/talkd/talkd.c
+++ b/talkd/talkd.c
@@ -41,6 +41,7 @@ time_t max_request_ttl = MAX_LIFE;
 
 char *acl_file;
 char *hostname;
+idx_t hostname_len;
 
 const char args_doc[] = "";
 const char doc[] = "Talk daemon, using service `ntalk'.";
@@ -131,6 +132,7 @@ talkd_init (void)
       syslog (LOG_ERR, "Cannot determine my hostname: %m");
       exit (EXIT_FAILURE);
     }
+  hostname_len = strlen (hostname);
 }
 
 time_t last_msg_time;

